### DISCUSSION TITLE
Possible to use drivers and access scene data? And regarding programming...

### DISCUSSION THREAD

**Discussion by festivities** [bnpr/Malt#265] [Q&A] [answered]
Hiya, I was wondering if it's possible to use drivers so I could for example, get bone rotation data from my rig or perhaps utilize two empties forming a vector? I tried right clicking as I would expect to do in EEVEE and it doesn't seem like this is the way to do it. May I ask how if it's possible?

![image](https://user-images.githubusercontent.com/77230051/156908170-94a193be-04b1-4874-950b-69ae6776753d.png)

Lastly, if it is possible is there any way I could do it in pure code (.mesh.glsl)?

Thank you!

### **EDIT:**
Oh, I also moved from the latest Release version to the Development branch pre-release version and now the include directive for _NPR_Pipeline.glsl_ doesn't work anymore.

`#include "Pipelines/NPR_Pipeline.glsl"`

I looked at the files and apparently the Pipeline is now located outside the Shaders folder - how do I approach fully programming a shader now? Thanks again.

**Comment by pragma37** [comment, marked as answer]
There's a limitation in Blender that doesn't allow to retrieve the relative path (and therefore adding drivers) for some types of node properties:
https://developer.blender.org/T51096

For now you can override the property from the material panel and add the driver there:
![Recording 2022-03-06 at 20 29 28](https://user-images.githubusercontent.com/36610999/156939075-2d38c84c-b1b9-46aa-b7cc-cd1cf633a25e.gif)

I have thought about adding the option to setup Malt properties with Python callbacks, but I'm on the fence since this would prevent some scene loading performance optimizations.

In any case, it's not possible to setup drivers directly from the shader code.

---

> Oh, I also moved from the latest Release version to the Development branch pre-release version and now the include directive for NPR_Pipeline.glsl doesn't work anymore.

The shader API has been redesignes and some new features have made fully writing shaders in GLSL a bit harder.
I think the best workflow for 1.0 is to move the shader code to plain functions and use those functions as nodes.

So where you used to write something like:
```glsl
#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform vec3 diffuse_color = vec3(1.0,0.1,0.1);

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    vec3 diffuse = diffuse_color * get_diffuse_half();
    PO.color.rgb = ambient_color + diffuse;
}
```
now it would be:
```glsl
/*META
    @ambient_color: default=vec3(0.1,0.1,0.1);
    @diffuse_color: default=vec3(1.0,0.1,0.1);
*/
vec3 basic_shading(vec3 ambient_color, vec3 diffuse_color)
{
    vec3 diffuse = diffuse_color * diffuse_half_shading();
    return ambient_color + diffuse;
}
```
and load the file either as a local library or as a [plugin library](https://github.com/bnpr/Malt/tree/Development/plugins/PluginExample):
![imagen](https://user-images.githubusercontent.com/36610999/156939696-28d647fa-8714-4a60-b7ee-1322e9899bbc.png)
